File: Capacity To Ship Packages Within D Days - GFG/capacity-to-ship-packages-within-d-days.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
//User function Template for C++

class Solution {
  public:
    int leastWeightCapacity(int weights[], int n, int days) {
         int  high = 0 , low = 1;
        for(int i = 0 ; i < n ; i++){
            high += weights[i];
            if(weights[i] > low)
                low = weights[i];
        }

        // int low = *max_element(weights.begin(),weights.end());

        while(high >= low){
            int mid = (low+high) >> 1;

            int d = 1,sum = 0;
            for(int i = 0 ; i < n ; i++){
                sum += weights[i];
                if(sum > mid){
                    sum -= weights[i];
                    d++;
                    sum = weights[i];
                }
            }

            if(d <= days)
                high = mid-1;
            else
                low = mid+1;
        }

        return low;
    }
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        int N,D;
        cin>>N;
        
        int arr[N];
        for(int i=0; i<N; i++)
            cin>>arr[i];
        cin>>D;

        Solution ob;
        cout << ob.leastWeightCapacity(arr,N,D) << endl;
    }
    return 0;
}
// } Driver Code Ends

File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.py
class Solution:
    def shiping_days(self, weights, limit):
        days = 1
        count = 0

        for i in weights:
            count += i

            if count > limit:
                days += 1
                count = i
        
        return days

    def shipWithinDays(self, weights: List[int], days: int) -> int:
        n = len(weights)
        low = max(weights)
        high = sum(weights)

        while low <= high:
            mid = (low + high) // 2

            ship_day = self.shiping_days(weights, mid)

            if ship_day <= days:
                high = mid - 1
            else:
                low = mid + 1
        
        return low
